Добавить сваггер для эндпойнтов авторизации (#77)

birthdaysgift opened the pull request:
Перед тем как приступить к реализации эндпойнтов авторизации, необходимо продумать их архитектуру и создать контракты.

Было принято решение реализовать упрощённую реализацию OAuth своими силами, чтобы лучше разобраться в работе протокола. Описание концепций, которые мы планируем реализовать находится на отдельной вики [странице](https://github.com/week-password/wisher/wiki/%D0%90%D0%BD%D0%B0%D0%BB%D0%B8%D1%82%D0%B8%D0%BA%D0%B0.-%D0%90%D0%B2%D1%82%D0%BE%D1%80%D0%B8%D0%B7%D0%B0%D1%86%D0%B8%D1%8F) (если эта ссылка, по какой-то причине отвалилась, то смотри комментарии к тикету #77 на гитхабе).

В рамках этой задачи необходимо продумать архитектуру эндпойнтов авторизации и добавить определения сваггера.

fowesn commented on src/auth/fields.py:
```diff
@@ -0,0 +1,23 @@
+class Password(SecretStrField):
+    """Password value field for User model."""
+
+    _PUNCTUATION: Final = r"""!"#$%&'()*+,-./:;<=>?@^_`{}|~\[\]\\"""
```
поскольку эти символы последовательны в таблице юникода, можно использовать их специальные обозначения через дефис `\x21 -x7E`, как [тут](https://github.com/week-password/wlss-frontend/blob/develop/src/app/core/validators/password.validators.ts#L4))

birthdaysgift replied:
Круто! Не знал про это, надо попробовать)

PS. Оставил ещё один [коммент](https://github.com/week-password/wlss-backend/pull/85#discussion_r1261311548) по этой же теме в другом посте.

Rey-Lane commented on src/auth/fields.py:
```diff
@@ -0,0 +1,23 @@
+    """Password value field for User model."""
+
+    _PUNCTUATION: Final = r"""!"#$%&'()*+,-./:;<=>?@^_`{}|~\[\]\\"""
+    _RU_LETTERS: Final = r"АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯабвгдеёжзийклмнопрстуфхцчшщъыьэюя"  # noqa: RUF001
```
А нельзя это сделать в виде `[а-яА-Я]`? 
Это ведь только в REGEX используется

birthdaysgift replied:
Изначально решил явно перечислить все буквы, потому что я не уверен, что все движки регулярок поддерживают синтаксис с русскими буквами типа `[а-я]`. Потому что например  движок может не знать, что это именно русский алфавит, а не белорусский, или какой-нибудь другой алфавит использующий кириллицу.

Но @fowesn подсказала фичу благдаря которой можно перечислять символы по их кодам в таблице кодировки, т.е. вместо `[А-Яа-я]` в пайтоне можно написать `[\u0410-\u044F]`. Читаемость у такого решения конечно хуже, но зато движок точно не должен ошибиться при компиляции этой регулярки.

Чтобы увеличить читаемость можно изобрести небольшой велосипед, который возвращал бы utf-8 код для символа, тогда регулярка могла бы выглядеть как-то так: `r"[unicode('А')-unicode('я')]"`. Скорее всего так и сделаем.

С другой стороны возможно движок регулярки под капотом делает то же самое что и наша `unicode()`, а это значит, что выражение типа `[А-я]` может работать из коробки... хммм... В общем надо проверить)

birthdaysgift replied:
> С другой стороны возможно движок регулярки под капотом делает то же самое что и наша `unicode()`, а это значит, что выражение типа `[А-я]` может работать из коробки... хммм... В общем надо проверить)

Формального подтверждения этому я нигде не нашёл, ни в доке, ни на сторонних ресурсах. Зато провел практический эксперимент, и выяснил, что это будет норм работать.

Вот скрипт который я написал для проверки работы этих регулярок и он отрабатывает без ошибок:
```python
import re


for char in "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯабвгдеёжзийклмнопрстуфхцчшщъыьэюя":
    if re.match(r"[А-яЁё]", char) is None:
        raise Exception(f"Char didn't match: {char}")


# ASCII символы с 32-го по 126 включительно
for char in """ !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~""":
    if re.match(r"[ -~]", char) is None:
        raise Exception(f"Char didn't match: {char}")
```

а это значит, что в регулярках можно обозначать диапазоны, типа `[А-я]` и регулярка будет их компилить как диапазон по кодам символов.

birthdaysgift replied:
Но в связи с грядущими изменениями в модели (см [пр](https://github.com/week-password/wlss-wiki/pull/2) в вики-репозитории), конкретно для пароля скорее всего никаких регулярок не будет. В остальных местах, например для имени пользователя, я использую эту фичу вместо перечисления каждой буквы в регулярке.

birthdaysgift replied:
Убрал ограничения символов в регулярке пароля. В остальных местах, где это необходимо, использовал диапазоны символов

Rey-Lane commented on src/auth/routes.py:
```diff
@@ -0,0 +1,226 @@
+    "/users/{user_id}/tokens/{device_id}",
+    responses={
+        status.HTTP_204_NO_CONTENT: {
+            "description": "All device user's tokens are removed.",
```
Может быть чуть попроще сказать?
Например: `All user tokens removed from device`

birthdaysgift replied:
Да, что-то я тут не по-русски как-то написал)

Думаю, можно даже ещё проще: `All device tokens are removed.`

birthdaysgift replied:
Поправил

Rey-Lane commented on src/auth/routes.py:
```diff
@@ -0,0 +1,226 @@
+    "/users/{user_id}/tokens",
+    responses={
+        status.HTTP_204_NO_CONTENT: {
+            "description": "All user's tokens are removed. User has been signed out from all devices.",
```
У меня всегда вызывает сомнения эта `'s`. Думаю, что её можно просто выкинуть и ничего не изменится. Кажется, что выражение `user tokens` вполне корректно

birthdaysgift replied:
Ну в данном случае этот апостроф не так плох, как в предыдущем) но по большому счёту да, мы можем от него избавиться, `user tokens` вполне норм

birthdaysgift replied:
Поправил

Rey-Lane commented on src/shared/fields.py:
```diff
@@ -0,0 +1,143 @@
+    """String field which value should be secret and not exposed in logging accidentally.
+
+    We duplicated all StrField functionality here instead of using multiple inheritance in order to avoid
+    method resolution order issues which may come with multiple inheritance.
```
Они `may come` или уже `came` пару раз?)

birthdaysgift replied:
Пока что только `may come`) Просто не очень люблю множественное наследование, поэтому стараюсь свести его к минимуму там, где возможно)

Rey-Lane commented on pyproject.toml:
```diff
@@ -20,6 +20,7 @@ alembic = {extras = ["tz"], version = "1.10.4"}
 asyncpg = "0.27.0"  # asynchronous postgresql driver used by sqlalchemy
 fastapi = {extras = ["all"], version = "0.95.0" }  # we need "all" at least for uvicorn
 minio = "7.1.14"
+overrides = "7.3.1"
```
А для чего эта штука нужна?
Как она в текущем коде используется?

birthdaysgift replied:
Вообще она использовалась, но видимо во время рефакторинга её куда-то унесло) я перепроверю

birthdaysgift replied:
Вернул декоратор `@override` в код. Используется он для того, чтобы явно обозначить, что метод переопределяет метод базового класса, а также он следит за тем, что метод обозначенный этим декоратором **действительно переопределяет** метод базового класса.

Это может спасти нас от ситуации, когда мы наследуемся от класса какой-либо библиотеки, переопределяем там метод, а после обновления библиотеки, этот метод из класса библиотеки пропадает. В таком случае этот декоратор нас сразу же предупредит, что на самом деле наш метод больше ничего не переопределяет, т.к. в базовом классе такого метода больше нет.

Также у библиотечки ещё есть интересные фишки, например, сравнение сигнатуры переопределяемых методов, или например автоматическое копирование докстринга из метода базового класса в переоределенный метод в дочернем классе (в рантайме). А также возможность объявить в базовом классе методы обязательные для переопределения.

Rey-Lane reviewed (approved):
![image](https://github.com/week-password/wlss-backend/assets/65604613/7b51312a-7a0a-4f18-82be-5881f7f3e513)

birthdaysgift committed:
Добавить сваггер для эндпойнтов авторизации (#77)

Перед тем как приступить к реализации эндпойнтов авторизации, необходимо продумать их архитектуру и создать контракты.

Было принято решение реализовать упрощённую реализацию OAuth своими силами, чтобы лучше разобраться в работе протокола. Описание концепций, которые мы планируем реализовать находится на отдельной вики [странице](https://github.com/week-password/wisher/wiki/%D0%90%D0%BD%D0%B0%D0%BB%D0%B8%D1%82%D0%B8%D0%BA%D0%B0.-%D0%90%D0%B2%D1%82%D0%BE%D1%80%D0%B8%D0%B7%D0%B0%D1%86%D0%B8%D1%8F) (если эта ссылка, по какой-то причине отвалилась, то смотри комментарии к тикету #77 на гитхабе).

В рамках этой задачи необходимо продумать архитектуру эндпойнтов авторизации и добавить определения сваггера.
